File: tests/end-to-end/stack_set/golang/stack.go
package main

import (
	"fmt"

	cdk "github.com/aws/aws-cdk-go/awscdk/v2"
	iam "github.com/aws/aws-cdk-go/awscdk/v2/awsiam"
	"github.com/aws/constructs-go/constructs/v10"
	"github.com/aws/jsii-runtime-go"
)

type StackSetStackProps struct {
	cdk.StackProps
	ModuleName *string
	RoleName *string
	RolePath *string
}

/// Deploy required components for StackSet custom resources in this region.  Lambda ARN is exported as StackSetCustomResource
type StackSetStack struct {
	cdk.Stack
}

func NewStackSetStack(scope constructs.Construct, id string, props *StackSetStackProps) *StackSetStack {
	var sprops cdk.StackProps
	if props != nil {
		sprops = props.StackProps
	}
	stack := cdk.NewStack(scope, &id, &sprops)

	useRoleName := !props.RoleName == jsii.String("")

	useRolePath := !props.RolePath == jsii.String("")

	iam.NewCfnRole(
		stack,
		jsii.String("StackSetResourceRole"),
		&iam.CfnRoleProps{
			RoleName: ifCondition(
				useRoleName,
				props.RoleName,
				nil,
			),
			Path: ifCondition(
				useRolePath,
				props.RolePath,
				jsii.String("/"),
			),
			AssumeRolePolicyDocument: map[string]interface{} {
				"Version": jsii.String("2012-10-17"),
				"Statement": &[]interface{}{
					map[string]interface{} {
						"Effect": jsii.String("Allow"),
						"Principal": map[string]interface{} {
							"Service": jsii.String("lambda.amazonaws.com"),
						},
						"Action": jsii.String("sts:AssumeRole"),
					},
				},
			},
			Policies: &[]interface{}{
				&PolicyProperty{
					PolicyName: jsii.String("IAMPassRolePermissions"),
					PolicyDocument: map[string]interface{} {
						"Version": jsii.String("2012-10-17"),
						"Statement": &[]interface{}{
							map[string]interface{} {
								"Effect": jsii.String("Allow"),
								"Action": jsii.String("iam:PassRole"),
								"Resource": jsii.String("*"),
							},
						},
					},
				},
				&PolicyProperty{
					PolicyName: jsii.String("CloudFormationPermissions"),
					PolicyDocument: map[string]interface{} {
						"Version": jsii.String("2012-10-17"),
						"Statement": &[]interface{}{
							map[string]interface{} {
								"Effect": jsii.String("Allow"),
								"Action": jsii.String("cloudformation:*"),
								"Resource": jsii.String("*"),
							},
						},
					},
				},
				&PolicyProperty{
					PolicyName: jsii.String("LambdaPermissions"),
					PolicyDocument: map[string]interface{} {
						"Version": jsii.String("2012-10-17"),
						"Statement": &[]interface{}{
							map[string]interface{} {
								"Effect": jsii.String("Allow"),
								"Action": jsii.String("logs:CreateLogGroup"),
								"Resource": &[]interface{}{
									jsii.String(fmt.Sprintf("arn:aws:logs:%v:%v:*", stack.Region(), stack.Account())),
								},
							},
							map[string]interface{} {
								"Effect": jsii.String("Allow"),
								"Action": &[]interface{}{
									jsii.String("logs:CreateLogStream"),
									jsii.String("logs:PutLogEvents"),
								},
								"Resource": &[]interface{}{
									jsii.String(fmt.Sprintf("arn:aws:logs:%v:%v:log-group:/aws/lambda/*", stack.Region(), stack.Account())),
								},
							},
						},
					},
				},
				&PolicyProperty{
					PolicyName: jsii.String("S3Permissions"),
					PolicyDocument: map[string]interface{} {
						"Version": jsii.String("2012-10-17"),
						"Statement": &[]interface{}{
							map[string]interface{} {
								"Effect": jsii.String("Allow"),
								"Action": &[]interface{}{
									jsii.String("s3:Get*"),
									jsii.String("s3:List*"),
								},
								"Resource": jsii.String("*"),
							},
						},
					},
				},
			},
		},
	)

	return &StackSetStack{
		Stack: stack,
	}
}

/// ifCondition is a helper function that replicates the ternary
/// operator that can be found in other languages. It is conceptually
/// equivalent to writing `cond ? whenTrue : whenFalse`, meaning it
/// returns `whenTrue` if `cond` is `true`, and `whenFalse` otherwise.
func ifCondition[T any](cond bool, whenTrue T, whenFalse T) T {
	if cond {
		return whenTrue
	}
	return whenFalse
}

File: tests/end-to-end/stack_set/python/stack.py
from aws_cdk import Stack
import aws_cdk as cdk
import aws_cdk.aws_iam as iam
from constructs import Construct

"""
  Deploy required components for StackSet custom resources in this region.  Lambda ARN is exported as StackSetCustomResource
"""
class StackSetStack(Stack):
  def __init__(self, scope: Construct, construct_id: str, **kwargs) -> None:
    super().__init__(scope, construct_id, **kwargs)

    # Applying default props
    props = {
      'moduleName': kwargs.get('moduleName', 'lambda_function'),
      'roleName': kwargs.get('roleName', ''),
      'rolePath': kwargs.get('rolePath', ''),
    }

    # Conditions
    use_role_name = not (props['roleName'] == '')
    use_role_path = not (props['rolePath'] == '')

    # Resources
    stackSetResourceRole = iam.CfnRole(self, 'StackSetResourceRole',
          role_name = props['roleName'] if use_role_name else None,
          path = props['rolePath'] if use_role_path else '/',
          assume_role_policy_document = {
            'Version': '2012-10-17',
            'Statement': [
              {
                'Effect': 'Allow',
                'Principal': {
                  'Service': 'lambda.amazonaws.com',
                },
                'Action': 'sts:AssumeRole',
              },
            ],
          },
          policies = [
            {
              'policyName': 'IAMPassRolePermissions',
              'policyDocument': {
                'Version': '2012-10-17',
                'Statement': [
                  {
                    'Effect': 'Allow',
                    'Action': 'iam:PassRole',
                    'Resource': '*',
                  },
                ],
              },
            },
            {
              'policyName': 'CloudFormationPermissions',
              'policyDocument': {
                'Version': '2012-10-17',
                'Statement': [
                  {
                    'Effect': 'Allow',
                    'Action': 'cloudformation:*',
                    'Resource': '*',
                  },
                ],
              },
            },
            {
              'policyName': 'LambdaPermissions',
              'policyDocument': {
                'Version': '2012-10-17',
                'Statement': [
                  {
                    'Effect': 'Allow',
                    'Action': 'logs:CreateLogGroup',
                    'Resource': [
                      f"""arn:aws:logs:{self.region}:{self.account}:*""",
                    ],
                  },
                  {
                    'Effect': 'Allow',
                    'Action': [
                      'logs:CreateLogStream',
                      'logs:PutLogEvents',
                    ],
                    'Resource': [
                      f"""arn:aws:logs:{self.region}:{self.account}:log-group:/aws/lambda/*""",
                    ],
                  },
                ],
              },
            },
            {
              'policyName': 'S3Permissions',
              'policyDocument': {
                'Version': '2012-10-17',
                'Statement': [
                  {
                    'Effect': 'Allow',
                    'Action': [
                      's3:Get*',
                      's3:List*',
                    ],
                    'Resource': '*',
                  },
                ],
              },
            },
          ],
        )


